[DevTools] feat: show changed hooks names in the Profiler tab

piotrski opened the pull request:
<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory.

  Before submitting a pull request, please make sure the following is done:

  1. Fork [the repository](https://github.com/facebook/react) and create your branch from `main`.
  2. Run `yarn` in the repository root.
  3. If you've fixed a bug or added code that should be tested, add tests!
  4. Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
  5. Run `yarn test --prod` to test in the production environment. It supports the same options as `yarn test`.
  6. If you need a debugger, run `yarn test --debug --watch TestName`, open `chrome://inspect`, and press "Inspect".
  7. Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
  8. Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
  9. Run the [Flow](https://flowtype.org/) type checks (`yarn flow`).
  10. If you haven't already, complete the CLA.

  Learn more about contributing: https://reactjs.org/docs/how-to-contribute.html
-->

## Summary

This PR adds support for displaying the names of changed hooks directly in the Profiler tab, making it easier to identify specific updates.

A `HookChangeSummary` component has been introduced to show these hook names, with a `displayMode` prop that toggles between `“compact”` for tooltips and `“detailed”` for more in-depth views. This keeps tooltip summaries concise while allowing for a full breakdown where needed.

This functionality also respects the `“Always parse hook names from source”` setting from the Component inspector, as it uses the same caching mechanism already in place for the Components tab. Additionally, even without hook names parsed, the Profiler will now display hook types (like `State`, `Callback`, etc.) based on data from `inspectedElement`.

To enable this across the DevTools, `InspectedElementContext` has been moved higher in the component tree, allowing it to be shared between the Profiler and Components tabs. This update allows hook name data to be reused across tabs without duplication.

Additionally, a `getAlreadyLoadedHookNames` helper function was added to efficiently access cached hook names, reducing the need for repeated fetching when displaying changes.

These changes improve the ability to track specific hook updates within the Profiler tab, making it clearer to see what’s changed.

### Before
Previously, the Profiler tab displayed only the IDs of changed hooks, as shown below:
<img width="350" alt="Screenshot 2024-11-01 at 12 02 21_cropped" src="https://github.com/user-attachments/assets/7a5f5f67-f1c8-4261-9ba3-1c76c9a88af3">

### After (without hook names parsed)
When hook names aren’t parsed, custom hooks and hook types are displayed based on the inspectedElement data:
<img width="350" alt="Screenshot 2024-11-01 at 12 03 09_cropped" src="https://github.com/user-attachments/assets/ed857a6d-e6ef-4e5b-982c-bf30c2d8a7e2">

### After (with hook names parsed)
Once hook names are fully parsed, the Profiler tab provides a complete breakdown of specific hooks that have changed:
<img width="350" alt="Screenshot 2024-11-01 at 12 03 14_cropped" src="https://github.com/user-attachments/assets/1ddfcc35-7474-4f4d-a084-f4e9f993a5bf">

This should resolve #21856 🎉

hoxyq commented:
Hey, thanks for upstreaming this, really appreciate any contributions to React DevTools.

I am currently on leave and will be back first week of December. This change will probably require some manual testing on my side, so I think its better to iterate on this once I am fully available.

hoxyq commented on packages/react-devtools-shared/src/devtools/views/Profiler/HookChangeSummary.js:
```diff
@@ -0,0 +1,185 @@
+    );
+    const store = useContext(StoreContext);
+
+    const [parseHookNamesOptimistic, setParseHookNamesOptimistic] =
+      useState(parseHookNames);
+
+    useEffect(() => {
+      setParseHookNamesOptimistic(parseHookNames);
+    }, [inspectedElement?.id, parseHookNames]);
```
This looks strange, you are keeping something from a global context in a local state, and then reconciliate it here? Why `parseHookNames` from `InspectedElementContext` is not enough?

piotrski replied:
This approach is something I borrowed from the Inspected Element Panel [here](https://github.com/facebook/react/blob/79ddf5b574db085104d917c24a964cbd5b824e09/packages/react-devtools-shared/src/devtools/views/Components/InspectedElementHooksTree.js#L57).

The idea behind the `parseHookNamesOptimistic` state is to keep the UI responsive. Since changing `parseHookNames` happens in a transition (which can suspend), this local state ensures the toggle feels immediate while the actual state updates in the background.

I considered making this logic reusable, by extracting `LoadHookNamesButton` into its own component, but I didn’t want to optimize prematurely. Do you think it’s worth revisiting?

hoxyq replied:
Oh, okay, we should probably refactor this to use `useOptimistic` in the future.

hoxyq commented on packages/react-devtools-shared/src/devtools/views/Profiler/HookChangeSummary.js:
```diff
@@ -0,0 +1,185 @@
+        .filter(Boolean);
+    }, [inspectedElement?.hooks, hooks]);
+
+    const hookParsingFailed = useMemo(
+      () => parseHookNames && hookNames === null,
+      [parseHookNames, hookNames],
+    );
```
You probably don't need `useMemo` for this one

hoxyq commented on packages/react-devtools-shared/src/devtools/views/Profiler/HookChangeSummary.js:
```diff
@@ -0,0 +1,185 @@
+import * as React from 'react';
```
This file is missing Flow typings, you can trigger it by adding a comment header, similar to other files:

```
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */
```

hoxyq commented on packages/react-devtools-shared/src/devtools/views/Profiler/HookChangeSummary.js:
```diff
@@ -0,0 +1,211 @@
+  if (hook.id !== null && hooksArray.includes(hook.id)) {
+    return true;
+  }
+  return (
+    hook.subHooks?.some(subHook => shouldKeepHook(subHook, hooksArray)) ?? false
+  );
```
```suggestion
  const subHooks = hook.subHooks;
  if (subHooks == null) {
    return false;
  }
  
  return subHooks.some(subHook => shouldKeepHook(subHook, hooksArray));
```

hoxyq commented on packages/react-devtools-shared/src/devtools/views/Profiler/HookChangeSummary.js:
```diff
@@ -0,0 +1,211 @@
+    return null;
+  }
+
+  if (hook.subHooks?.length > 0) {
+    const filteredSubHooks = hook.subHooks
+      .map(subHook => filterHooks(subHook, hooksArray))
+      .filter(Boolean);
+
+    return filteredSubHooks.length > 0
+      ? {...hook, subHooks: filteredSubHooks}
+      : {...hook};
+  }
+
+  return hook;
```
```suggestion
  const subHooks = hook.subHooks;
  if (subHooks == null) {
    return hook;
  }
  
  const filteredSubHooks = subHooks
      .map(subHook => filterHooks(subHook, hooksArray))
      .filter(Boolean);

  return filteredSubHooks.length > 0
      ? {...hook, subHooks: filteredSubHooks}
      :  hook;
```

hoxyq commented on packages/react-devtools-shared/src/devtools/views/Profiler/HookChangeSummary.js:
```diff
@@ -0,0 +1,211 @@
+        </span>
+        {hook.subHooks?.map((subHook, index) => (
+          <Hook
+            key={`${hook.id}-${index}`}
```
Do you really need an `index` here for constructing a key? `hook.id` and `index` actually have a correlation, because both of them depend on the fact when the hook was defined relatively to other hooks: in the end all of them are going to be stored in a linked list.

piotrski replied:
I remember it was a dirty fix to prevent stale hook info, but I couldn't reproduce it now, so I removed it.

hoxyq reviewed (commented):
Looks good overall, please address the comment that I've left.

I am planning to manually test this next week, and if everything works as expected, we can merge this

hoxyq commented:
@piotrski Are you planning to continue the work on this or are you okay with someone picking this up?

piotrski commented:
@hoxyq sorry, I was sick, and then I got caught up with other things and completely missed this. I’ll try to get it done this week since I finally have some time for OSS

hoxyq commented:
> @hoxyq sorry, I was sick, and then I got caught up with other things and completely missed this. I’ll try to get it done this week since I finally have some time for OSS

No worries, thank you!

piotrski commented:
@hoxyq I’ve addressed all the comments, let me know if anything else needs further adjustments. Looking forward to your feedback.
I’d like to add a diff of the hook values, but I didn’t want this PR to grow too much, so I think it’s better to do it in a separate PR. I’m not sure about the UI for that yet.
Also, hook name parsing is a bit flaky, sometimes it fails silently, but I haven’t been able to figure out why.

hoxyq commented:
> I’d like to add a diff of the hook values, but I didn’t want this PR to grow too much, so I think it’s better to do it in a separate PR.

Makes sense, that would be awesome. 

> Also, hook name parsing is a bit flaky, sometimes it fails silently, but I haven’t been able to figure out why.

Yeah, the implementation is flawed, and was based on a lot of assumptions about the source code, but that's what we have. This should not block you from landing this change, since you are just using it as an API.

hoxyq commented:
Hey, just circling back here to say that this is still in my queue.

In my scenario, I end up in a weird state where I can see the button for fetching hook names on the inspected element panel, but not on the profile tab. We should probably avoid doing this. I am not sure yet what is going wrong, though.

Will follow up here once I have more details.

piotrski commented:
@hoxyq thanks for the update! Happy to jump on a quick call if that helps speed up debugging, just let me know.

hoxyq reviewed (approved):
All good with changes in this diff, thanks for contributing. I am going to proceed with merging this.

Before the next release, I will try to find some time to fix the hook names parsing implementation, because right now it is mostly broken for Flow and extremely slow for larger apps. I don't have an approximal date for next release yet.

piotrski commented:
Thanks for merging! Yeah, I’ve noticed the hook name parsing can be a bit flaky too, especially in Next.js projects. If there's anything I can do to help with that, just let me know. I'm happy to help.
